cmd/allocator/main.go

```diff
@@ -380,11 +400,22 @@ func runGRPC(h *serviceHandler, grpcPort int) {
 
 	grpcServer := grpc.NewServer(h.getGRPCServerOptions()...)
 	pb.RegisterAllocationServiceServer(grpcServer, h)
+	grpc_health_v1.RegisterHealthServer(grpcServer, grpcHealth)
 
 	go func() {
+		go func() {
+			<-ctx.Done()
+			grpcServer.Stop()
```

Comment:
Should this be: https://pkg.go.dev/google.golang.org/grpc#Server.GracefulStop (or maybe used in lieu of a lot of this code?)

Which might help with your multi-context issue?

Reply:
I went back and forth on this, and I have a version in a branch where I did exactly that. It made things worse.

Let me just go back to the drawing board here, but I think it may involve rewriting a lot of the handling basically everywhere.

Reply:
But no, it doesn't solve the two-context issue. The problem is that here (and in httpserver/https), we use a context to signal shutdown, but the service handler still needs a forward context to get work done to shut down. Let me see if I can just rework the listen handlers more.

Reply:
Yeah I see your point. I think you are right that we'll want two contexts, but what we could do is something akin to:

```go
<- listenerCtx.Done()
// graceful shutdown both the http and grpc listeners. 
grpcServer.GracefulStop()
server.Shutdown()
cancelInfrormer() // cancel the worker + Informer context, since now we know we aren't accepting any more connections.
```

This might require pulling the grpc + http listeners into a larger struct that has a `Run(ctx)` and `Stop()` operation on it, to provide a larger data structure to operate on (and then it could take a context, and manage it's own context for the workers and informers).

Did that make sense?

Reply:
Hmm, we could do something similar to the runner slices and have shutdown slices, perhaps?

Reply:
> Hmm, we could do something similar to the runner slices and have shutdown slices, perhaps?

Seems like you are rebuilding Contexts 😄